oai_agents/common/teammates_collection.py

```diff
@@ -6,92 +6,128 @@
-            used_agents.update([tm[0] for tm in tms_prftg_scr])
-
-        elif teamtype == TeamType.RANDOM:
-            tms_prftg_scr = random.sample(available_agents, teammates_len)
-            all_teammates[teamtype].append([tm[0] for tm in tms_prftg_scr])
-            used_agents.update([tm[0] for tm in tms_prftg_scr])
-
-        elif teamtype == TeamType.ALL_MIX:
-            teammate_permutations = list(permutations(sorted_agents_perftag_score, teammates_len))
-            for tp in teammate_permutations:
-                all_teammates[teamtype].append([tm[0] for tm in tp])
-
-        elif teamtype == TeamType.SELF_PLAY:
-            assert agent is not None
-            all_teammates[teamtype].append([agent for _ in range(teammates_len)])
-
-        elif teamtype == TeamType.SELF_PLAY_HIGH:
-            assert agent is not None
-            high_p_agents = available_agents[:unseen_teammates_len]
-            agents_itself = [agent for _ in range(teammates_len - unseen_teammates_len)]
-            all_teammates[teamtype].append([tm[0] for tm in high_p_agents] + agents_itself)
-            used_agents.update([tm[0] for tm in high_p_agents])
-
-        elif teamtype == TeamType.SELF_PLAY_MEDIUM:
-            assert agent is not None
-            mean_score = (available_agents[0][2] + available_agents[-1][2]) / 2
-            mean_p_agents = sorted(available_agents, key=lambda x: abs(x[2] - mean_score))[:unseen_teammates_len]
-            agents_itself = [agent for _ in range(teammates_len - unseen_teammates_len)]
-            all_teammates[teamtype].append([tm[0] for tm in mean_p_agents] + agents_itself)
-            used_agents.update([tm[0] for tm in mean_p_agents])
-
-        elif teamtype == TeamType.SELF_PLAY_MIDDLE:
-            assert agent is not None
-            middle_index = len(available_agents) // 2
-            start_index_for_mid = middle_index - unseen_teammates_len // 2
-            end_index_for_mid = start_index_for_mid + unseen_teammates_len
-            mid_p_agents = available_agents[start_index_for_mid:end_index_for_mid]
-            agents_itself = [agent for _ in range(teammates_len - unseen_teammates_len)]
-            all_teammates[teamtype].append([tm[0] for tm in mid_p_agents] + agents_itself)
-            used_agents.update([tm[0] for tm in mid_p_agents])
-
-        elif teamtype == TeamType.SELF_PLAY_LOW:
-            assert agent is not None
-            low_p_agents = available_agents[-unseen_teammates_len:]
-            agents_itself = [agent for _ in range(teammates_len - unseen_teammates_len)]
-            all_teammates[teamtype].append([tm[0] for tm in low_p_agents] + agents_itself)
-            used_agents.update([tm[0] for tm in low_p_agents])
-
-
-    selected_agents = []
-    for teamtype in teamtypes:
-        selected_agents.extend(all_teammates[teamtype])
-
-    return all_teammates, selected_agents
+    entire_population_used = False
+
+    while True:
```

Comment:
We must only use all of the population if H, M, L are present at the same time. For example, If train_types are SPH, SPM we should not include low performance agents in the TC. We should only collect all the H and Ms.

Reply:
@avv-va Ok interesting. I may need to re-think this a little then. Two follow up questions:

1. if H and L are are the only two train types, can I just discard any agent checkpoint with an M performance tag? My hesitation is that right now, an agent checkpoint with a "medium" tag could be better than an agent checkpoint with a "high" tag. (.e.g SP_M2 > SP_H1). The perf tags don't necessarily dictate the TrainType team that the checkpoint ends up on.

2. Should the behavior be the same for any training method? (N-1-SP and FCP?)

Reply:
Hmm good questions, I think we might need to completely change the way we create teammates. 
Q1: I think we should not rely on the tags and instead do this: 

Since all the population is currently stored here and is sorted based on score: 
`sorted_agents_perftag_score = sorted(agents_perftag_score, key=lambda x: x[2], reverse=True)`
We can categorize them based on their score and not the tags. Since 3 checkpoints are saved per agent, that means that 1/3 are H, 1/3 are M, 1/3 are L. We can create a dict such that: 
```
agents_categorized = {
'H': [the first 1/3 of the agents],
'M': [the second 1/3 of the agents],
'L': [the last 1/3 of the agents]
}
```
Then each time we enter an if statement [here](https://github.com/HIRO-group/multiHRI/blob/2552629130f9cb672678747a6492a076221cdd7c/oai_agents/common/teammates_collection.py#L57) we can construct all the teammates instead of iterativing over multiple times until `entire_population_used=True`.

For example, if training types were SP types (SELF_PLAY_HIGH, SELF_PLAY_MEDIUM, SELF_PLAY_LOW) then we can create a TC collection by doing something like: 
```
for teamtype in teamtypes:
    if teamtype == TeamType.SELF_PLAY_HIGH:
        agents_itself = [agent for _ in range(teammates_len - unseen_teammates_len)] 
        all_teammates[teamtype] = [agents_itself+[x] for x in agents_categorized['H'] # You can assume unseen_teammates_len = 1 for now and later implement it for unseen_teammates_len>1
```
Notice how this time I don't use append [here](https://github.com/HIRO-group/multiHRI/blob/2552629130f9cb672678747a6492a076221cdd7c/oai_agents/common/teammates_collection.py#L61). That's because in one iteration all the SPH teammates are gonna be created. 

You can only implement this part ^ and forget about none-SP teamtypes such as HIGH_FIRST (as long as it doesn't break SP(), FCP_traditional(), SPN_1ADV_XSPCKP functions) for now but if you wanted to work on that too this is what I suggest:

We can do the same for non_SP_types as well and construct the teammates without needing to iterate over it. It's a bit trickier because for example in case of high_first in a 3 player setting meaning that your high_first should look like: [h1, h2], but you might only have 3 high performance agents in agents_categorized['H'] (i.e. agents_categorized['H']=[h1, h2, h3]). I think in this case it's okay to repeat (copy) agents in each category until it's len(agents_categorized['H'])%N-1 == 0. 
```
so agents_categorized['H']=[h1, h2, h3, h1] and then 
all_teammates[HIGH_FIRST] = [[h1 ,h2], [h3, h1]]
```

Reply:
Ok I think this makes sense to me but do you want me to keep the use of the entire population as optional? (i.e. Do I need [this argument](https://github.com/HIRO-group/multiHRI/blob/full-pop-in-tc/oai_agents/common/teammates_collection.py#L9C134-L9C155) or should we create the TC with the entire population every time?) 

If we always use the entire population, there will be restrictions on the size of the pop as it relates to the number of TrainTypes. I think [this check](https://github.com/HIRO-group/multiHRI/blob/full-pop-in-tc/oai_agents/common/teammates_collection.py#L22) accounts for it.

Reply:
Good point let's keep this [argument](https://github.com/HIRO-group/multiHRI/blob/full-pop-in-tc/oai_agents/common/teammates_collection.py#L9C134-L9C155) for now, we can change it later

Reply:
I updated teammates_collection.py to reflect this discussion. 
Note that right now, this assumes that number of train_types is less than or equal to 3, if more than 3 train types are provided only 1 set of teammates will be provided 
e.g. if `TrainTypes = [SPH, SPH, SPM]`
The TC will only have one list for `SPH`
`{layout_name: {'SPH' : [[team1], [team2]] }, 'SPM' : [...]}`

Also, note that I have some `NotImplementedError` exceptions in place. I can work on addressing those but I wanted to get this commit in as soon as possible